caddyhttp/httpserver/server.go

```diff
@@ -292,40 +285,99 @@ func (s *Server) Address() string {
 
 // Stop stops s gracefully (or forcefully after timeout) and
 // closes its listener.
-func (s *Server) Stop() (err error) {
-	s.Server.SetKeepAlivesEnabled(false)
-
-	if runtime.GOOS != "windows" {
-		// force connections to close after timeout
-		done := make(chan struct{})
-		go func() {
-			s.connWg.Done() // decrement our initial increment used as a barrier
-			s.connWg.Wait()
-			close(done)
-		}()
+func (s *Server) Stop() error {
+	if s.closed {
+		return errors.New("Server has been closed")
+	}
 
-		// Wait for remaining connections to finish or
-		// force them all to close after timeout
-		select {
-		case <-time.After(s.connTimeout):
-		case <-done:
+	// Make sure a listener was set
+	if s.listener != nil {
+		// Close the listener to stop all new connections
+		if err := s.listener.Close(); err != nil {
+			return err
 		}
 	}
 
-	// Close the listener now; this stops the server without delay
-	s.listenerMu.Lock()
-	if s.listener != nil {
-		err = s.listener.Close()
-		s.listener = nil
+	s.Server.SetKeepAlivesEnabled(false)
+	s.mu.Lock()
+	s.closed = true
+	s.mu.Unlock()
+
+	// force connections to close after timeout
+	wait := make(chan struct{})
+	go func() {
+		defer close(wait)
+		s.mu.Lock()
+		for c, st := range s.conns {
```

Comment:
Are there any race conditions with net/http?
Consider the following timeline:

1. Stop called
2. We got the lock for closing idle/new conns
3. Just before we close an idle/new conn, a new request arrive on that conn in net/http
4. s.Server.ConnState is blocked by the lock
5. We close all idle/new conn and release the lock
6. ConnState carry on and pass control to handler
7. Handler process the request but fail to send response because the conn is closed